src/routing/bootstrap.rs

```diff
@@ -336,70 +173,125 @@ impl<'a> State<'a> {
 
+        let mut relocate_payload = None;
         loop {
+            used_recipient.extend(recipients.iter());
+
             let (response, sender, dest_info) = self
                 .receive_join_response(genesis_key.as_ref(), relocate_payload.as_ref())
                 .await?;
 
             match response {
+                JoinResponse::Rejected(JoinRejectionReason::NodeNotReachable(addr)) => {
+                    error!(
+                        "Node cannot join the network since it is not externally reachable: {}",
+                        addr
+                    );
+                    return Err(Error::NodeNotReachable(addr));
+                }
+                JoinResponse::Rejected(JoinRejectionReason::JoinsDisallowed) => {
+                    error!("Network is set to not taking any new joining node, try join later.");
+                    return Err(Error::TryJoinLater);
+                }
                 JoinResponse::Approval {
                     section_auth,
                     genesis_key,
                     section_chain,
+                    ..
                 } => {
                     return Ok((
                         self.node,
                         Section::new(genesis_key, section_chain, section_auth)?,
                         self.backlog.into_iter().collect(),
                     ));
                 }
-                JoinResponse::Retry {
-                    section_auth,
-                    section_key: new_section_key,
-                } => {
-                    if new_section_key == section_key {
+                JoinResponse::Retry(section_auth) | JoinResponse::Redirect(section_auth) => {
+                    if section_auth.section_key() == section_key {
                         continue;
                     }
 
-                    if section_auth.prefix().matches(&self.node.name()) {
+                    // Ignore already used recipients
+                    let new_recipients: Vec<(XorName, SocketAddr)> = section_auth
+                        .elders
+                        .iter()
+                        .filter(|(name, addr)| !used_recipient.contains(&(**name, **addr)))
+                        .map(|(name, addr)| (*name, *addr))
+                        .collect();
+
+                    if new_recipients.is_empty() {
+                        debug!("Joining redirected to the same set of peers we already contacted - ignoring response");
+                    } else {
+                        info!(
+                            "Joining redirected to another set of peers: {:?}",
+                            new_recipients,
+                        );
+                    }
+
+                    let prefix = section_auth.prefix;
+
+                    // For the first section, using age random among 6 to 100 to avoid
+                    // relocating too many nodes at the same time.
+                    if prefix.is_empty() && self.node.age() < FIRST_SECTION_MIN_AGE {
```

Comment:
will relocate happen during the first section?
if so, then we cannot carry out this step when under relocation?

Reply:
we currently do this step when relocating, we grab the prefix from the GetSectionInfo request